sailfishusermanager.cpp

```diff
@@ -348,3 +382,118 @@ bool SailfishUserManager::hasAccessRights(uint uid_to_modify) {
 
     return true;
 }
+
+void SailfishUserManager::switchUser(uint uid)
+{
+    if (checkCallerUid() == SAILFISH_UNDEFINED_UID)
+        return;
+
+    if (m_switchUser) {
+        auto message = QStringLiteral("Already switching user");
+        qCWarning(lcSUM) << message;
+        sendErrorReply(QStringLiteral(SailfishUserManagerErrorBusy), message);
+        return;
+    }
+
+    bool uidFound = false;
+    struct group *grent = getgrnam(USER_GROUP);
+    if (grent) {
+        for (int i = 0; grent->gr_mem[i]; i++) {
+            struct passwd *pw = getpwnam(grent->gr_mem[i]);
+            if (pw && pw->pw_uid == uid) {
+                uidFound = true;
+                break;
+            }
+        }
+    }
+    if (!uidFound) {
+        auto message = QStringLiteral("User not found");
+        qCWarning(lcSUM) << message;
+        sendErrorReply(QDBusError::InvalidArgs, message);
+        return;
+    }
+
+    m_currentUid = currentUser();
+    if (m_currentUid == SAILFISH_UNDEFINED_UID)
+        return;
+
+    if (m_currentUid == uid) {
+        // No need to do anything
+        emit userSwitched(uid);
```

Comment:
I would argue that this is now doing something (so it's not what the comment says). Usually it's best to not to emit if things don't change but it might be good to say that caller did something wrong instead. So basically spec the interface so that is shouldn't be called with current uid.

Reply:
Now sends an error here